hypothesis-python/tests/cover/test_lookup_py37.py

```diff
@@ -0,0 +1,212 @@
+import sys
+
+import pytest
+
+from hypothesis import assume, given, infer
+
+# This test file is skippped on Python <= 3.6, where `list[Elem]` is always a
+# SyntaxError.  On Python 3.7 and 3.8, `from __future__ import annotations` means
+# that the syntax is supported; but the feature fails at runtime.  On Python
+# 3.9 and later, it should all work.
+#
+# For details, see https://www.python.org/dev/peps/pep-0585/
+
+if sys.version_info < (3, 9):
+    pytestmark = pytest.mark.xfail(
+        raises=Exception, strict=True, reason="Requires Python 3.9 (PEP 585) or later."
+    )
+
+
+class Elem:
+    pass
+
+
+class Value:
+    pass
+
+
+def check(t, ex):
+    assert isinstance(ex, t)
+    assert all(isinstance(e, Elem) for e in ex)
+    assume(ex)
+
+
+@given(x=infer)
+def test_resolving_standard_tuple1_as_generic(x: tuple[Elem]):
+    check(tuple, x)
+
+
+@given(x=infer)
+def test_resolving_standard_tuple2_as_generic(x: tuple[Elem, Elem]):
+    check(tuple, x)
+
+
+@given(x=infer)
+def test_resolving_standard_tuple_variadic_as_generic(x: tuple[Elem, ...]):
+    check(tuple, x)
+
+
+@given(x=infer)
+def test_resolving_standard_list_as_generic(x: list[Elem]):
+    check(list, x)
+
+
+@given(x=infer)
+def test_resolving_standard_dict_as_generic(x: dict[Elem, Value]):
+    check(dict, x)
+    assert all(isinstance(e, Value) for e in x.values())
+
+
+@given(x=infer)
+def test_resolving_standard_set_as_generic(x: set[Elem]):
+    check(set, x)
+
+
+@given(x=infer)
+def test_resolving_standard_frozenset_as_generic(x: frozenset[Elem]):
+    check(frozenset, x)
+
+
+@given(x=infer)
+def test_resolving_standard_defaultdict_as_generic(
+    x: collections.defaultdict[Elem, Value]
+):
+    check(collections.defaultdict, x)
+    assert all(isinstance(e, Value) for e in x.values())
+
+
+@given(x=infer)
+def test_resolving_standard_iterable_as_generic(x: collections.abc.Iterable[Elem]):
+    check(collections.abc.Iterable, x)
+
+
+@given(x=infer)
+def test_resolving_standard_iterator_as_generic(x: collections.abc.Iterator[Elem]):
+    check(collections.abc.Iterator, x)
+
+
+@given(x=infer)
+def test_resolving_standard_reversible_as_generic(x: collections.abc.Reversible[Elem]):
+    check(collections.abc.Reversible, x)
+
+
+@given(x=infer)
+def test_resolving_standard_container_as_generic(x: collections.abc.Container[Elem]):
+    check(collections.abc.Container, x)
+
+
+@given(x=infer)
+def test_resolving_standard_collection_as_generic(x: collections.abc.Collection[Elem]):
+    check(collections.abc.Collection, x)
+
+
+@given(x=infer)
+def test_resolving_standard_callable_ellipsis(x: collections.abc.Callable[..., Elem]):
+    assert isinstance(x, collections.abc.Callable)
+    assert callable(x)
+    # ... implies *args, **kwargs; as would any argument types
+    assert isinstance(x(), Elem)
+    assert isinstance(x(1, 2, 3, a=4, b=5, c=6), Elem)
+
+
+@given(x=infer)
+def test_resolving_standard_callable_no_args(x: collections.abc.Callable[[], Elem]):
```

Comment:
What about callables like `Callable[[int], Elem]`?

Reply:
Good question - because of the problem of argument names, we only distinguish "no arguments" and "`*args, **kwargs`" when resolving `Callable`.